python/lsst/ip/isr/calibType.py

```diff
@@ -0,0 +1,630 @@
+
+        Raises
+        ------
+        NotImplementedError
+            This needs to be implemented by subclasses for each
+            calibration type.
+        """
+        raise NotImplementedError("Must be implemented by subclass.")
+
+    @classmethod
+    def fromDict(cls, dictionary):
+        """Construct a calibration from a dictionary of properties.
+
+        Must be implemented by the specific calibration subclasses.
+
+        Parameters
+        ----------
+        dictionary : `dict`
+            Dictionary of properties.
+
+        Returns
+        ------
+        calib : `lsst.ip.isr.CalibType`
+            Constructed calibration.
+
+        Raises
+        ------
+        NotImplementedError :
+            Raised if not implemented.
+        """
+        raise NotImplementedError("Must be implemented by subclass.")
+
+    def toDict(self):
+        """Return a dictionary containing the calibration properties.
+
+        The dictionary should be able to be round-tripped through
+        `fromDict`.
+
+        Returns
+        -------
+        dictionary : `dict`
+            Dictionary of properties.
+
+        Raises
+        ------
+        NotImplementedError :
+            Raised if not implemented.
+        """
+        raise NotImplementedError("Must be implemented by subclass.")
+
+    @classmethod
+    def fromTable(cls, tableList):
+        """Construct a calibration from a dictionary of properties.
+
+        Must be implemented by the specific calibration subclasses.
+
+        Parameters
+        ----------
+        tableList : `list` [`lsst.afw.table.Table`]
+            List of tables of properties.
+
+        Returns
+        ------
+        calib : `lsst.ip.isr.CalibType`
+            Constructed calibration.
+
+        Raises
+        ------
+        NotImplementedError :
+            Raised if not implemented.
+        """
+        raise NotImplementedError("Must be implemented by subclass.")
+
+    def toTable(self):
+        """Return a list of tables containing the calibration properties.
+
+        The table list should be able to be round-tripped through
+        `fromDict`.
+
+        Returns
+        -------
+        tableList : `list` [`lsst.afw.table.Table`]
+            List of tables of properties.
+
+        Raises
+        ------
+        NotImplementedError :
+            Raised if not implemented.
+        """
+        raise NotImplementedError("Must be implemented by subclass.")
+
+    def validate(self, other=None):
+        """Validate that this calibration is defined and can be used.
+
+        Parameters
+        ----------
+        other : object, optional
+            Thing to validate against.
+
+        Returns
+        -------
+        valid : `bool`
+            Returns true if the calibration is valid and appropriate.
+        """
+        return False
+
+    def apply(self, target):
+        """Method to apply the calibration to the target object.
```

Comment:
So for Defects this would take an afwExposure and update the mask?

Reply:
That's the idea that I had.  I don't know that it will be populated for every type (such as the IsrProvenance), but if there's a simple obvious way to apply the calibration, it would be good to have them use the same method name.